keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,386 @@
+
+If flavorFungibility is not nil and `WhenCanBorrow` is `TryNextFlavor`, we will try to borrow after all flavors have been consiedered.
+
+If flavorFungibility is not nil and `WhenCanPreempt` is `TryNextFlavor`, we will try to preempt after all flavors have been consiedered.
+
+If both `WhenCanPreempt` and `WhenCanBorrow` are `true`, we will first try borrowing before preempting.
+
+For example, if cluster queue has 2 resource groups and workload has 1 podSet as the following:
+
+```
+...
+  - coveredResources: ["cpu", "memory"]
+    flavors:
+    - name: "default-flavor1"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+    - name: "default-flavor2"
+      resources:
+      - name: "cpu"
+        nominalQuota: 3 
+      - name: "memory"
+        nominalQuota: 600Mi 
+  - coveredResources: ["gpu"]
+    flavors:
+    - name: "vendor1"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+    - name: "vendor2"
+      resources:
+      - name: "gpu"
+        nominalQuota: 9  
+---  
+...
+  podSets:
+  - count: 3
+    spec:
+      containers:
+      - ...
+        resources:
+          requests:
+            cpu: "1"
+            memory: 200Mi
+            gpu: 1
+```
+
+We will first try `default-flavor1` for cpu and memory resources. If `default-flavor1` doesn't fit, we try preempt in `default-flavor1`. If preemptor return preempt fail, we try to allocate resource in `default-flavor2`.
+
+We will pass preemptor to flavorAssigner and let flavorAssigner try to call public interface of preemptor to get candidate for certain podSet in current flavor if podSet can not find enough resources in current flavor.
+
+##### Advantages
+
+##### Disadvantages
+
+#### Plan B
+
+```
+type FlavorFungibility string
+
+const (
+  Borrow   FlavorFungibility = "Borrow"
+  Preempt  FlavorFungibility = "Preempt"
+  TryNextFlavor FlavorFungibility = "TryNextFlavor"
+)
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+  // +kubebuilder:validation:UniqueItems=true
+  // +kubebuilder:validation:Enum="Borrow,Preempt,TryNextFlavor"
+  FlavorFungibility []FlavorFungibility `json:"flavorFungibility"`
+}
+```
+
+Length of FlavorFungibility should be either 0 or 3, this means we don't allow users to omit any of these three actions. If FlavorFungibility is empty, we set FlavorFungibility as `Borrow,TryNextFlavor,Preempt`. Actions after `TryNextFlavor` will be executed after all flavors are considered, and actions before `TryNextFlavor` will be executed before next flavor be considered.
```

Comment:
I like the idea that all the actions should be listed. However, I'm not aware of any k8s API that looks like this.

I started a thread on slack https://kubernetes.slack.com/archives/CJUQN3E4T/p1686161408595329

Are you sure all permutations are valid? For example, if we try to Borrow before Preempting, there could be priority inversion, similar to this: https://github.com/kubernetes-sigs/kueue/pull/763#issuecomment-1551831935

Somehow `TryNextFlavor` is special. Should it be something like the following?

```yaml
beforeTryingNextFlavor: ["Preempt", "Borrow"]
```

Reply:
Currently we will borrow unused quota from other cluster queues before we try preempting. 
Do you mean we will change this behavior? 
If we can say that preemption always takes precedence over borrowing, I think plan A is better.